app/code/community/Bolt/Boltpay/controllers/ShippingController.php

```diff
@@ -152,45 +162,38 @@ public function prefetchEstimateAction()
             return;
         }
 
-        $shippingAddressOriginal = $quote->getShippingAddress()->getData();
+        $shippingAddress = array_intersect_key(
+            $quote->getShippingAddress()->getData(),
+            array('city'=>'','region'=>'','region_id'=>'','postcode'=>'','country_id'=>'')
+        );
 
-        $cacheIdentifier = $this->getPrefetchCacheIdentifier($quote, $shippingAddressOriginal);
-        $addressCacheKey = $this->getAddressCacheKey($cacheIdentifier);
+        $cachedIdentifier = $this->getEstimateCacheIdentifier($quote, $shippingAddress);
+        $estimate = $this->_cache->load($cachedIdentifier);
 
-        if ($serialized = $this->_cache->load($addressCacheKey)) {
-            $addressData = unserialize($serialized);
+        if ($estimate) {
+            $addressData = $shippingAddress;
         } else {
-            $geoLocationAddress = $this->getGeoIpAddress();
-            $geoLocationAddress = $this->cleanEmptyAddressField($geoLocationAddress);
-
-            // ----------^_^----------- //
-            $shippingAddress = array(
-                'city'       => @($shippingAddressOriginal['city']),
-                'region'     => @($shippingAddressOriginal['region']),
-                'region_id'  => @($shippingAddressOriginal['region_id']),
-                'postcode'   => @($shippingAddressOriginal['postcode']),
-                'country_id' => @($shippingAddressOriginal['country_id']),
-            );
-            unset($shippingAddressOriginal);
-
-            $addressData = $this->mergeAddressData($geoLocationAddress, $shippingAddress);
-
-            if(isset($addressData['postcode'])) {
-                $cacheIdentifier = $this->getPrefetchCacheIdentifier($quote, $addressData);
-                $this->saveAddressCache($addressData, $cacheIdentifier);
-
-                $quote->getShippingAddress()->addData($addressData);
-                $quote->getBillingAddress()->addData($addressData);
-
-                try {
-                    /** @var Bolt_Boltpay_Model_ShippingAndTax $shippingAndTaxModel */
-                    $shippingAndTaxModel = Mage::getModel('boltpay/shippingAndTax');
-                    $estimateResponse = $shippingAndTaxModel->getShippingAndTaxEstimate($quote);
-
-                    $this->cacheShippingAndTaxEstimate($estimateResponse, $cacheIdentifier);
-                } catch (Exception $e) {
-                    $estimateResponse = null;
-                }
+            $addressData = count($shippingAddress) === 5 ? $shippingAddress : array_merge(array_filter($this->getGeoIpAddress()), array_filter($shippingAddress));
```

Comment:
Fortunately for us these lines of code 

https://github.com/BoltApp/bolt-magento1/blob/6698900b99735fba696faa98b25e9719a55047c9/app/code/community/Bolt/Boltpay/controllers/ShippingController.php#L165-L168

take all of the magic out of `count($shippingAddress) === 5` and keep us from having to resort to an unnecessary, code obfuscating "helper" method.  The array can only have 5 entries at most and the entries must be of the defined keys `array('city'=>'','region'=>'','region_id'=>'','postcode'=>'','country_id'=>'')`.

Absolutely, I can rename and variables and write it so that is a little less magic looking and more self-describing.